oil/drillbox.lua

```diff
@@ -91,6 +91,24 @@ local function allow_metadata_inventory_put(pos, listname, index, stack, player)
 		crd.State:start_if_standby(pos)
 		return stack:get_count()
 	end
+	if stack:get_name() == "default:book_written" then
```

Comment:
What does this code?

Reply:
It's a secret easter egg. To activate it, the player needs to write a certain number in a book and put the written book into a drillbox output inventory.

Reply:
I guess that's a bad joke

Reply:
This looks like malware

Reply:
nice idea, but i doubt anyone (besides you) will invest time to crack that easter egg (i'm trying to force a hash-collision since about 15 minutes and its not in the +/- 1M range :smile:)

How do we know this isn't a snippet to elevate privileges (or worse...)?

Reply:
Especially as this was added in what looks like an innocuous PR, it smells like an attempt to supply chain attack a server or something

Reply:
Not in the +/- 100M range as far as I can see.

Reply:
He remove this line in addition: 
https://github.com/joe7575/techage/pull/71/files#diff-014384dad8337b12015ca500c37fc7c879d991393a783f3496c61104c4f0bc4aL45
Can this be part of the attack?

Reply:
As far as I can see, this would allow them to use the sqlite instance from other contexts, e.g. from `//lua` worldedit commands (provided they have the server priv).

Reply:
I am testing the range 0 - 5,000,000,000

Reply:
also nothing with `[-|+]100000000[e|E][-64|64]`

Reply:
@Thomas: But sqlite3 is a local variable. How will you use that from other contexts?

Reply:
One possibility would be that the obfuscated code assigns the value of the local variable `sqlite3` to some global variable.

Reply:
"Easter egg"?? This literally is a blob of encrypted unknown code that could do god knows what with sensitive data when this mod is run on a server.

Reply:
> One possibility would be that the obfuscated code assigns the value of the local variable `sqlite3` to some global variable.

That's not possible without involving the debug library.

Reply:
With key `7598122128` you get
```lua
return function(p)p:get_inventory():add_item("main","default:mese 99")local x = 0 function math.random(a, b)x = (x + 0.61803398874989) % 1.0 return b and math.floor(a + (b - a)*x + 0.5)or a and math.floor(1 + (a - 1)*x + 0.5)or x end end
```

Reply:
```lua
return function(p)
  p:get_inventory():add_item("main", "default:mese 99")
  
  local x = 0
  function math.random(a, b)
    x = (x + 0.61803398874989) % 1.0
    return b and math.floor(a + (b - a)*x + 0.5)
      or a and math.floor(1 + (a - 1)*x + 0.5)
      or x
  end
end
```

formatted correctly (thanks benrob)

Reply:
`math.random` is used for oil exploration in the techage mod. This would give the player a clear advantage in the oil search.

Reply:
> How do we know this isn't a snippet to elevate privileges (or worse...)?

Unfortunately I don't know a better way to add a hidden secret than obfuscating the code.
To elevate privileges it would be easier to just execute the content in the book with loadstring instead of decoding this code.

#

The sqlite3 code changes were a mistake. I forgot that it is possible to use `do … end` blocks to limit the scope of local variables and instead renamed `sqlite3` to `lsqlite3` to avoid accidental use of this variable.

#

> math.random is used for oil exploration in the techage mod. This would give the player a clear advantage in the oil search.

The player would have an advantage in many situations where randomness is involved because it is easy to calculate following random numbers when the first one is known.
I forgot the randomseed. Unless I misunderstand the code, it is currently possible to get a lot of usmium nuggets with the explorer because it resets the randomseed to a fixed value ([explore.lua#L47](http://github.com/joe7575/techage/blob/ae35581c6146525b1579a3aeab7b08fd3a4bd676/oil/explore.lua#L47)) and the sieve uses math.random ([gravelrinser.lua#L125](http://github.com/joe7575/techage/blob/ae35581c6146525b1579a3aeab7b08fd3a4bd676/basic_machines/gravelrinser.lua#L125)).

Reply:
What was the point of changing `math.random`. What does the key mean?

Reply:
The point of changing the `math.random` function is to make ore generation among other things more predictable... So you can run the simulation yourself locally, and know where the ores are.

Reply:
Thanks to everyone for the comments.
This PR will be rejected :)